src/Data/Singletons/Deriving/Show.hs

```diff
@@ -0,0 +1,140 @@
+                    , id_meths = [ (showsPrecName, UFunction clauses) ] }
+
+mk_showsPrec :: DsMonad q => [DCon] -> q [DClause]
+mk_showsPrec cons = do
+    p <- newUniqueName "p" -- The precedence argument (not always used)
+    mapM (mk_showsPrec_clause p) cons
+
+mk_showsPrec_clause :: forall q. DsMonad q => Name -> DCon -> q DClause
+mk_showsPrec_clause p (DCon _ _ con_name con_fields _) = go con_fields
+  where
+    go :: DConFields -> q DClause
+
+    -- No fields: print just the constructor name, with no parentheses
+    go (DNormalC _ []) = return $
+      DClause [DWildPa, DConPa con_name []] $
+        DVarE showStringName `DAppE` dStringE (parenInfixConName con_name "")
+
+    -- Infix constructors have special Show treatment.
+    go (DNormalC True [_, _]) = do
+      argL <- newUniqueName "argL"
+      argR <- newUniqueName "argR"
+      fi <- fromMaybe defaultFixity <$> reifyFixityWithLocals con_name
+      let con_prec = case fi of Fixity prec _ -> prec
+          op_name  = nameBase con_name
+          infixOpE = DAppE (DVarE showStringName) . dStringE $
+                       if isInfixDataCon op_name
+                          then " "  ++ op_name ++ " "
+                          -- Make sure to handle infix data constructors
+                          -- like (Int `Foo` Int)
+                          else " `" ++ op_name ++ "` "
+      return $ DClause [DVarPa p, DConPa con_name [DVarPa argL, DVarPa argR]] $
+        (DVarE showParenName `DAppE` (DVarE gtName `DAppE` DVarE p
+                                                   `DAppE` dIntegerE con_prec))
+          `DAppE` (DVarE composeName
+                     `DAppE` showsPrecE (con_prec + 1) argL
+                     `DAppE` (DVarE composeName
+                                `DAppE` infixOpE
+                                `DAppE` showsPrecE (con_prec + 1) argR))
+
+    go (DNormalC _ tys) = do
+      args <- mapM (const $ newUniqueName "arg") tys
+      let show_args     = map (showsPrecE appPrec1) args
+          composed_args = foldr1 (\v q -> DVarE composeName
+                                           `DAppE` v
+                                           `DAppE` (DVarE composeName
+                                                     `DAppE` DVarE showSpaceName
+                                                     `DAppE` q)) show_args
+          named_args = DVarE composeName
+                         `DAppE` (DVarE showStringName
+                                   `DAppE` dStringE (parenInfixConName con_name " "))
+                         `DAppE` composed_args
+      return $ DClause [DVarPa p, DConPa con_name $ map DVarPa args] $
+        DVarE showParenName
+          `DAppE` (DVarE gtName `DAppE` DVarE p `DAppE` dIntegerE appPrec)
+          `DAppE` named_args
+
+    -- A record constructor with no fields? This probably won't ever happen,
+    -- but just to be safe, we'll act as if it were a DNormalC.
+    go (DRecC []) = go (DNormalC False [])
```

Comment:
Ah yes. I thought that GHC turned `data Foo = Foo {}` into a `NormalC`, but that's only if you _reify_ it. `[d| data Foo = Foo {} |]` does get turned into `RecC`, as I just experimentally confirmed. I'll edit the comment accordingly.